File: app/schemas/insurance.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import date
from enum import Enum


class InsuranceTypeEnum(str, Enum):
    """Enum for different insurance types."""
    LIFE = "LIFE"
    DISABILITY = "DISABILITY"
    CRITICAL_ILLNESS = "CRITICAL_ILLNESS"
    LONG_TERM_CARE = "LONG_TERM_CARE"
    HEALTH = "HEALTH"
    HOME = "HOME"
    AUTO = "AUTO"
    OTHER = "OTHER"


class InsurancePolicyBase(BaseModel):
    """Base insurance policy schema with common attributes."""
    name: str
    policy_number: Optional[str] = None
    insurance_type: InsuranceTypeEnum
    provider: Optional[str] = None
    coverage_amount: float = Field(0.0, ge=0.0)
    premium_amount: float = Field(0.0, ge=0.0)
    premium_payment_frequency: Optional[str] = "annual"
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    is_term: bool = True
    is_taxable_benefit: bool = False
    notes: Optional[str] = None


class InsurancePolicyCreate(InsurancePolicyBase):
    """Schema for creating a new insurance policy."""
    family_member_id: int


class InsurancePolicyUpdate(BaseModel):
    """Schema for updating insurance policy information."""
    name: Optional[str] = None
    policy_number: Optional[str] = None
    insurance_type: Optional[InsuranceTypeEnum] = None
    provider: Optional[str] = None
    coverage_amount: Optional[float] = Field(None, ge=0.0)
    premium_amount: Optional[float] = Field(None, ge=0.0)
    premium_payment_frequency: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    is_term: Optional[bool] = None
    is_taxable_benefit: Optional[bool] = None
    notes: Optional[str] = None
    family_member_id: Optional[int] = None


class InsurancePolicy(InsurancePolicyBase):
    """Schema for insurance policy information returned to clients."""
    id: int
    user_id: int
    family_member_id: int

    class Config:
        from_attributes = True


class InsurancePolicyList(BaseModel):
    """Schema for a list of insurance policies."""
    policies: List[InsurancePolicy] 

File: frontend/src/api/insurance.ts
import { api } from './api';

export interface InsurancePolicy {
  id: number;
  name: string;
  policy_number: string | null;
  insurance_type: InsuranceTypeEnum;
  provider: string | null;
  coverage_amount: number;
  premium_amount: number;
  premium_payment_frequency: string | null;
  start_date: string | null;
  end_date: string | null;
  is_term: boolean;
  is_taxable_benefit: boolean;
  notes: string | null;
  family_member_id: number;
}

export enum InsuranceTypeEnum {
  LIFE = "LIFE",
  DISABILITY = "DISABILITY",
  CRITICAL_ILLNESS = "CRITICAL_ILLNESS",
  LONG_TERM_CARE = "LONG_TERM_CARE",
  HEALTH = "HEALTH",
  HOME = "HOME",
  AUTO = "AUTO",
  OTHER = "OTHER"
}

export interface InsurancePolicyCreate {
  name: string;
  policy_number?: string | null;
  insurance_type: InsuranceTypeEnum;
  provider?: string | null;
  coverage_amount?: number;
  premium_amount?: number;
  premium_payment_frequency?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  is_term?: boolean;
  is_taxable_benefit?: boolean;
  notes?: string | null;
  family_member_id: number;
}

export interface InsurancePolicyUpdate {
  name?: string;
  policy_number?: string | null;
  insurance_type?: InsuranceTypeEnum;
  provider?: string | null;
  coverage_amount?: number;
  premium_amount?: number;
  premium_payment_frequency?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  is_term?: boolean;
  is_taxable_benefit?: boolean;
  notes?: string | null;
  family_member_id?: number;
}

export const insuranceApi = {
  /**
   * Get all insurance policies, optionally filtered by family member and type
   */
  getAll: async (familyMemberId?: number, insuranceType?: string): Promise<InsurancePolicy[]> => {
    let params = '';
    
    if (familyMemberId || insuranceType) {
      params = '?';
      if (familyMemberId) {
        params += `family_member_id=${familyMemberId}`;
      }
      if (insuranceType) {
        params += familyMemberId ? `&insurance_type=${insuranceType}` : `insurance_type=${insuranceType}`;
      }
    }
    
    return api.get(`/insurance-policies${params}`);
  },

  /**
   * Get a single insurance policy by ID
   */
  getById: async (id: number): Promise<InsurancePolicy> => {
    return api.get(`/insurance-policies/${id}`);
  },

  /**
   * Create a new insurance policy
   */
  create: async (data: InsurancePolicyCreate): Promise<InsurancePolicy> => {
    return api.post('/insurance-policies', data);
  },

  /**
   * Update an existing insurance policy
   */
  update: async (id: number, data: InsurancePolicyUpdate): Promise<InsurancePolicy> => {
    return api.put(`/insurance-policies/${id}`, data);
  },

  /**
   * Delete an insurance policy
   */
  delete: async (id: number): Promise<void> => {
    return api.delete(`/insurance-policies/${id}`);
  }
}; 